apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -87,50 +92,137 @@ public override int GetHashCode()
+                    {
+                        match.Value.Release();
+                        _multiClientPool.Remove(match.Key);
+                        _multiClientPool.Add(match.Key, match.Value);
+                    }
+                }
+            }
+        }
+
+        private class SpannerClientCreator : IComparable<SpannerClientCreator>, IComparable
+        {
+            private  Lazy<Task<SpannerClient>> _creationTask;
+            private int _refCount = 0;
+            private ClientCredentialKey _key;
+
+            public SpannerClientCreator(ClientCredentialKey key)
+            {
+                _key = key;
+                _creationTask = new Lazy<Task<SpannerClient>>(AcquireClientImplAsync);
+            }
+
+            public bool MatchesClient(SpannerClient client) => _creationTask.IsValueCreated &&
+                ReferenceEquals(_creationTask.Value.ResultWithUnwrappedExceptions(), client);
+
+            private async Task<ChannelCredentials> CreateDefaultChannelCredentialsAsync()
+            {
+                var appDefaultCredentials = await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
+                if (appDefaultCredentials.IsCreateScopedRequired)
+                {
+                    appDefaultCredentials = appDefaultCredentials.CreateScoped(SpannerClient.DefaultScopes);
                 }
+                return appDefaultCredentials.ToChannelCredentials();
+            }
+
+            private async Task<SpannerClient> AcquireClientImplAsync()
+            {
+                var endpoint = _key.Endpoint ?? SpannerClient.DefaultEndpoint;
+                ChannelCredentials channelCredentials;
+
+                if (_key.Credential == null)
+                {
+                    channelCredentials = await CreateDefaultChannelCredentialsAsync().ConfigureAwait(false);
+                }
+                else
+                {
+                    channelCredentials = _key.Credential.ToChannelCredentials();
+                }
+
+                var channel = new Channel(
+                    endpoint.Host,
+                    endpoint.Port,
+                    channelCredentials);
+                return SpannerClient.Create(channel);
+            }
 
-                return _client;
+            public Task<SpannerClient> AcquireClientAsync()
+            {
+                Interlocked.Increment(ref _refCount);
+                return _creationTask.Value;
+            }
+
+            public void Release()
+            {
+                Interlocked.Decrement(ref _refCount);
+            }
+
+            /// <inheritdoc />
+            public int CompareTo(SpannerClientCreator other)
+            {
+                GaxPreconditions.CheckNotNull(other, nameof(other));
+
+                if (ReferenceEquals(this, other))
+                {
+                    return 0;
+                }
+                var result = _refCount.CompareTo(other._refCount);
```

Comment:
right -- prioritylist will only ever process one prioritychange at a time.  In that processing is where compareto is called.
so an item could theoretically be handling multiple concurrent addRefs which would each result in a prioritychange event.
The prioritylist will be 'eventually consistent' in that the last prioritychange event will result in the proper sort order.  Intermediate prioritychange events could see refcnt changes from other events, but that's ok.